### ISSUE TITLE
Cluster Allocation Explain changed substantially in 5.2

### ISSUE THREAD

**Issue by Mpdreamz** [elastic/elasticsearch-net#2573]
https://www.elastic.co/guide/en/elasticsearch/reference/5.2/cluster-allocation-explain.html

```5.2
{
  "index" : "project",
  "shard" : 0,
  "primary" : true,
  "current_state" : "started",
  "current_node" : {
    "id" : "04uXR0X2TaObVA8Mm-fhHA",
    "name" : "readonly-node-06180c",
    "transport_address" : "127.0.0.1:9300",
    "attributes" : {
      "testingcluster" : "true"
    },
    "weight_ranking" : 1
  },
  "can_remain_on_current_node" : "yes",
  "can_rebalance_cluster" : "yes",
  "can_rebalance_to_other_node" : "no",
  "rebalance_explanation" : "cannot rebalance as no target node exists that can both allocate this shard and improve the cluster balance"
}
```

```.5.3
{
  "shard" : {
    "index" : "project",
    "index_uuid" : "byQtcKrFQceVZFiL8GLssQ",
    "id" : 0,
    "primary" : true
  },
  "assigned" : true,
  "assigned_node_id" : "pEda2shWT1Sw_wb-Yuiknw",
  "shard_state_fetch_pending" : false,
  "nodes" : {
    "pEda2shWT1Sw_wb-Yuiknw" : {
      "node_name" : "readonly-node-21da5b",
      "node_attributes" : {
        "testingcluster" : "true"
      },
      "store" : {
        "shard_copy" : "AVAILABLE"
      },
      "final_decision" : "ALREADY_ASSIGNED",
      "final_explanation" : "the shard is already assigned to this node",
      "weight" : 0.0,
      "decisions" : [
        {
          "decider" : "max_retry",
          "decision" : "YES",
          "explanation" : "shard has no previous failures"
        },
        {
          "decider" : "replica_after_primary_active",
          "decision" : "YES",
          "explanation" : "shard is primary and can be allocated"
        },
        {
          "decider" : "enable",
          "decision" : "YES",
          "explanation" : "all allocations are allowed"
        },
        {
          "decider" : "node_version",
          "decision" : "YES",
          "explanation" : "target node version [5.1.1] is the same or newer than source node version [5.1.1]"
        },
        {
          "decider" : "snapshot_in_progress",
          "decision" : "YES",
          "explanation" : "no snapshots are currently running"
        },
        {
          "decider" : "filter",
          "decision" : "YES",
          "explanation" : "node passes include/exclude/require filters"
        },
        {
          "decider" : "same_shard",
          "decision" : "NO",
          "explanation" : "the shard cannot be allocated on the same node id [pEda2shWT1Sw_wb-Yuiknw] on which it already exists"
        },
        {
          "decider" : "disk_threshold",
          "decision" : "YES",
          "explanation" : "there is only a single data node present"
        },
        {
          "decider" : "throttling",
          "decision" : "YES",
          "explanation" : "below shard recovery limit of outgoing: [0 < 2] incoming: [0 < 2]"
        },
        {
          "decider" : "shards_limit",
          "decision" : "YES",
          "explanation" : "total shard limits are disabled: [index: -1, cluster: -1] <= 0"
        },
        {
          "decider" : "awareness",
          "decision" : "YES",
          "explanation" : "allocation awareness is not enabled"
        }
      ]
    }
  }
}
```

Which puts us in a bit of a bind on how to support this moving forward. cc @elastic/microsoft

**Comment by russcam** [comment]
This is an API level breaking change 😞 

Looking at the examples on https://www.elastic.co/guide/en/elasticsearch/reference/5.2/cluster-allocation-explain.html, it looks like we may be able to handle at least some of the structural differences by mapping to our existing type e.g. construct a `ShardAllocationExplanation` type, etc.

**Comment by gmarz** [comment]
Closed via #2589
